src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/HttpApplication/RequestHttpApplicationFeature.cs

```diff
@@ -0,0 +1,89 @@
+        _previous = previousEnd;
+        _context = context;
+    }
+
+    public RequestNotification CurrentNotification { get; set; }
+
+    public bool IsPostNotification { get; set; }
+
+    public bool IsEnded => _previous.IsEnded;
+
+    public HttpApplication Application { get; }
+
+    public ValueTask RaiseEventAsync(ApplicationEvent @event)
+    {
+        if (IsEnded)
+        {
+            return ValueTask.CompletedTask;
+        }
+
+        (CurrentNotification, IsPostNotification) = @event switch
+        {
+            ApplicationEvent.BeginRequest => (RequestNotification.BeginRequest, false),
+            ApplicationEvent.AuthenticateRequest => (RequestNotification.AuthenticateRequest, false),
+            ApplicationEvent.PostAuthenticateRequest => (RequestNotification.AuthenticateRequest, true),
+            ApplicationEvent.AuthorizeRequest => (RequestNotification.AuthorizeRequest, false),
+            ApplicationEvent.PostAuthorizeRequest => (RequestNotification.AuthorizeRequest, true),
+            ApplicationEvent.ResolveRequestCache => (RequestNotification.ResolveRequestCache, false),
+            ApplicationEvent.PostResolveRequestCache => (RequestNotification.ResolveRequestCache, true),
+            ApplicationEvent.MapRequestHandler => (RequestNotification.MapRequestHandler, false),
+            ApplicationEvent.PostMapRequestHandler => (RequestNotification.MapRequestHandler, true),
+            ApplicationEvent.AcquireRequestState => (RequestNotification.AcquireRequestState, false),
+            ApplicationEvent.PostAcquireRequestState => (RequestNotification.AcquireRequestState, true),
+            ApplicationEvent.PreRequestHandlerExecute => (RequestNotification.PreExecuteRequestHandler, false),
+            ApplicationEvent.PostRequestHandlerExecute => (RequestNotification.ExecuteRequestHandler, true),
+            ApplicationEvent.ReleaseRequestState => (RequestNotification.ReleaseRequestState, false),
+            ApplicationEvent.PostReleaseRequestState => (RequestNotification.ReleaseRequestState, true),
+            ApplicationEvent.UpdateRequestCache => (RequestNotification.UpdateRequestCache, false),
+            ApplicationEvent.PostUpdateRequestCache => (RequestNotification.UpdateRequestCache, true),
+            ApplicationEvent.LogRequest => (RequestNotification.LogRequest, false),
+            ApplicationEvent.PostLogRequest => (RequestNotification.LogRequest, true),
+            ApplicationEvent.EndRequest => (RequestNotification.EndRequest, false),
+
+            // Remaining events just continue using the existing notifications
+            _ => (CurrentNotification, IsPostNotification),
+        };
+
+        Application.InvokeEvent(@event);
+
+        return ValueTask.CompletedTask;
+    }
+
+    async Task IHttpResponseEndFeature.EndAsync()
+    {
+        // Prevent subsequent calls from going through this pathway again
+        _context.Features.Set(NoopEndFeature.Instance);
```

Comment:
This only works if the next time they fetch the feature again, not if they're already holding onto an existing one? Would it be better to track state in this instance and reset it later if you're pooling these.

Reply:
good call. I rework it so it track that state inside